src/training/stylegan.py

```diff
@@ -0,0 +1,154 @@
+
+            reconstucted = self.model(style_label_embedds, style_embeds)
+            reconstucted_loss = self.l1(style_batch, reconstucted)
+
+            reconstucted_style_embedds = self.style_embedder(reconstucted)
+            cycle = self.model(style_label_embedds, reconstucted_style_embedds)
+            cycle_loss = self.l1(style_batch, cycle)
+
+            loss = self.coef_ocr * ocr_loss + self.coef_perceptual * perceptual_loss \
+                 + self.coef_cycle * cycle_loss + self.coef_recon * reconstucted_loss
+
+            loss.backward()
+            self.optimizer.step()
+
+            self.logger.log_train(
+                losses={
+                    'ocr_loss': ocr_loss.item(),
+                    'perceptual_loss': perceptual_loss.item(),
+                    'full_loss': loss.item(),
+                    'cycle': cycle_loss.item(),
+                    'reconstruction': reconstucted_loss.item()},
+                images={
+                    'style': style_batch,
+                    'content': content_batch,
+                    'result': res,
+                    'recognized': word_images})
+
+    def validate(self, epoch):
+        self.model.eval()
+        self.content_embedder.eval()
+        self.style_embedder.eval()
+
+        for style_batch, content_batch, label_batch in self.val_dataloader:
+            if max(len(label) for label in label_batch) > 25:
+                continue
+            style_batch = style_batch.to(self.device)
+            content_batch = content_batch.to(self.device)
+            style_embeds = self.style_embedder(style_batch)
+            content_embeds = self.content_embedder(content_batch)
+
+            res = self.model(content_embeds, style_embeds)
+            ocr_loss = self.ocr_loss(res, label_batch)
+            perceptual_loss = self.perceptual_loss(style_batch, res)
+            loss = self.coef_ocr * ocr_loss + self.coef_perceptual * perceptual_loss
+
+            self.logger.log_val(
+                losses={
+                    'ocr_loss': ocr_loss.item(),
+                    'perceptual_loss': perceptual_loss.item(),
+                    'full_loss': loss.item()},
+                images={
+                    'style': style_batch,
+                    'content': content_batch,
+                    'result': res})
+
+        avg_losses, _ = self.logger.end_val()
+        self.storage.save(epoch,
+                          {'model': self.model,
+                           'content_embedder': self.content_embedder,
+                           'style_embedder': self.style_embedder,
+                           'optimizer': self.optimizer,
+                           'scheduler': self.scheduler},
+                          avg_losses['full_loss'])
+
+    def run(self):
+        for epoch in range(self.total_epochs):
+            self.train()
+            with torch.no_grad():
+                self.validate(epoch)
+            if self.scheduler is not None:
```

Comment:
I am still puzzled -- why do we need it? Why do we do step here?

Reply:
Usually a scheduler step is done each epoch

Reply:
what is the question? didnt get it